File: client/lib/rpc/packing_types.dart
import 'package:batufo/rpc/generated/message_bus.pb.dart';

class Point {
  final int x;
  final int y;

  // 0000 0000 0000 0000 1111 1111 1111 1111
  static const mask = 0xffff;
  static const shift = 16;

  const Point(this.x, this.y);

  PackedPoint pack() {
    final xy = x << shift | y;
    return PackedPoint()..xy = xy;
  }

  factory Point.unpack(PackedPoint data) {
    final xy = data.xy;
    final y = xy & mask;
    final x = xy >> shift;
    return Point(x, y);
  }

  @override
  String toString() {
    return 'Point { x: $x, y: $y }';
  }
}

int _doubleToInt(double value, int factor) => (value * factor).floor();

DoubleTwoDecimals packTwoDecimals(double value) {
  return DoubleTwoDecimals()..value = _doubleToInt(value, 100);
}

DoubleFourDecimals packFourDecimals(double value) {
  return DoubleFourDecimals()..value = _doubleToInt(value, 10000);
}

double unpackTwoDecimals(DoubleTwoDecimals data) {
  return data.value / 1E2;
}

double unpackFourDecimals(DoubleFourDecimals data) {
  return data.value / 1E4;
}

///
/// The max doubles for x/y that can be packed is +-32767
class FractionalPoint {
  final double x;
  final double y;

  static const mask = 0xffff;
  static const signMask = 0x8000;
  static const unsignMask = 0x7FFF;

  static const factor = 100;

  const FractionalPoint(this.x, this.y);

  PackedFractionalPoint pack({int factor = FractionalPoint.factor}) {
    final sintX = (x * factor).floor();
    final sintY = (y * factor).floor();

    final signX = sintX < 0 ? signMask : 0;
    final signY = sintY < 0 ? signMask : 0;
    final intX = sintX.abs() | signX;
    final intY = sintY.abs() | signY;
    final xy = intX << 16 | intY;
    return PackedFractionalPoint()..xy = xy;
  }

  factory FractionalPoint.unpack(
    PackedFractionalPoint data, {
    int factor = FractionalPoint.factor,
  }) {
    final xy = data.xy;
    final intY = xy & mask;
    final intX = xy >> 16;
    final xSigned = (intX & signMask) > 0;
    final ySigned = (intY & signMask) > 0;
    final x = xSigned ? -(intX & unsignMask) : intX;
    final y = ySigned ? -(intY & unsignMask) : intY;
    return FractionalPoint(x / factor, y / factor);
  }

  String toString() {
    return 'FractionalPoint{x: $x, y: $y}';
  }
}

File: backend/packages/server/src/rpc/packing-types.ts
import { DoubleTwoDecimals, PackedFractionalPoint, PackedPoint } from '../generated/message_bus_pb'

const doubleToInt = (value: number, factor: number) =>
  Math.floor(value * factor)

export function packTwoDecimals(value: number): DoubleTwoDecimals {
  const d2d = new DoubleTwoDecimals()
  d2d.setValue(doubleToInt(value, 1e2))
  return d2d
}

export function packFourDecimals(value: number): DoubleTwoDecimals {
  const d4d = new DoubleTwoDecimals()
  d4d.setValue(doubleToInt(value, 1e4))
  return d4d
}

export function unpackTwoDecimals(data: DoubleTwoDecimals): number {
  return data.getValue() / 1e2
}

export function unpackFourDecimals(data: DoubleTwoDecimals): number {
  return data.getValue() / 1e4
}

export class Point {
  constructor(readonly x: number, readonly y: number) {}

  // 0000 0000 0000 0000 1111 1111 1111 1111
  static mask = 0xffff
  static shift = 16

  pack(): PackedPoint {
    const xy = (this.x << Point.shift) | this.y
    const point = new PackedPoint()
    point.setXy(xy)
    return point
  }

  static unpack(data: PackedPoint): Point {
    const xy = data.getXy()
    const y = xy & Point.mask
    const x = xy >> Point.shift
    return new Point(x, y)
  }

  toString(): string {
    return `Point { x: ${this.x}, y: ${this.y} }`
  }
}

export class FractionalPoint {
  static mask = 0xffff
  static signMask = 0x8000
  static unsignMask = 0x7fff

  static factor = 1e2

  constructor(readonly x: number, readonly y: number) {}

  pack(): PackedFractionalPoint {
    const sintX = Math.floor(this.x * FractionalPoint.factor)
    const sintY = Math.floor(this.y * FractionalPoint.factor)

    const signX = sintX < 0 ? FractionalPoint.signMask : 0
    const signY = sintY < 0 ? FractionalPoint.signMask : 0
    const intX = Math.abs(sintX) | signX
    const intY = Math.abs(sintY) | signY
    const xy = (intX << 16) | intY
    const fp = new PackedFractionalPoint()
    fp.setXy(xy)
    return fp
  }

  static unpack(data: PackedFractionalPoint): FractionalPoint {
    const xy = data.getXy()
    const intY = xy & FractionalPoint.mask
    const intX = xy >> 16
    const xSigned = (intX & FractionalPoint.signMask) > 0
    const ySigned = (intY & FractionalPoint.signMask) > 0
    const x = xSigned ? -(intX & FractionalPoint.unsignMask) : intX
    const y = ySigned ? -(intY & FractionalPoint.unsignMask) : intY
    return new FractionalPoint(
      x / FractionalPoint.factor,
      y / FractionalPoint.factor
    )
  }
}
